### PR TITLE
Dev dice loss

### PR DESCRIPTION
**Pull request by apyskir**
*dice loss
*fix in steps/pytorch.validation.py to free memory at epoch end (from @taraspiotr )
*fix of loaders

### CODE REVIEW

File: loaders.py [hunk 1]
```diff
@@ -436,7 +436,7 @@ def get_datagen(self, X, y, train_mode, loader_params):
                                    train_mode=False,
                                    image_augment=None,
                                    image_augment_with_target=None,
-                                   mask_transform=None,
+                                   mask_transform=self.mask_transform,
```
Comment (jakubczakon) [inline loaders.py]: @apyskir is this intentional? Was it causing trouble when valid_datagen was used for validation during training? Does it mean `train_mode` may be confusing?

File: models.py [hunk 2]
```diff
@@ -280,19 +280,38 @@ def __get_distance_weights(d, w1, w0, sigma):
...
-    if torch.cuda.is_available():
-        C = C.cuda()
+def __get_size_weights(sizes, C):
     size_weights = C / sizes
     size_weights[sizes == 1] = 1
     return size_weights
 
 
-def get_loss_params(w0, sigma):
+def get_loss_params(w0, sigma, imsize):
```
Comment (jakubczakon) [inline models.py]: @apyskir It seems inconsistent to build tensors here but w1 tensor is build somewhere else. Either one or the other. I think that one should simply pass constants `w0` ,`sigma`,`C`  to the `multiclass_weighted_segmentation_loss` and create all the tensors in the same function. You could have a function create_weight_variables or something if you like to (which I think is not a bad idea) but just do all that in one place.

File: models.py [hunk 3]
```diff
@@ -280,19 +280,38 @@ def __get_distance_weights(d, w1, w0, sigma):
...
         w0 = w0.cuda()
         sigma = sigma.cuda()
-    return {'w0': w0, 'sigma': sigma}
+        C = C.cuda()
+    return {'w0': w0, 'sigma': sigma, 'C': C}
+
+
+def mixed_dice_cross_entropy_loss(output, target, dice_weight, ce_weight, ce_loss=None):
+    dice_target = target[:, 0, :, :].long()
+    ce_target = target
```
Comment (jakubczakon) [inline models.py]: @apyskir I kinda get the idea of having ce and dice targets separate but it confuses me. Those are in fact the very same target and it would make it clearer if it was in fact just the `target` passed around

File: models.py [hunk 4]
```diff
@@ -253,7 +256,7 @@ def callbacks_unet(callbacks_config):
                    ])
 
 
-def multiclass_weighted_segmentation_loss(output, target, w0, sigma):
+def multiclass_weighted_segmentation_loss(output, target, w0, sigma, C):
```
Comment (jakubczakon) [inline models.py]: @apyskir this is in fact multiclass_weighted_cross_entropy. The whole idea of segmentation_loss (which may actually be entirely flawed when you can simply sum stuff) was to have one function that handles both bce and dice. So I would change this name

File: models.py [hunk 5]
```diff
@@ -263,13 +266,10 @@ def multiclass_weighted_segmentation_loss(output, target, w0, sigma):
     w1 = Variable(torch.ones(distances.size()), requires_grad=False)  # TODO: fix it to handle class imbalance
     if torch.cuda.is_available():
         w1 = w1.cuda()
-    size_weights = __get_size_weights(sizes)
+    size_weights = __get_size_weights(sizes, C)
```
Comment (jakubczakon) [inline models.py]: @apyskir any reason why those are __get... . In case of methods I sort of know what it means (and it's way more than just differntiating between private and public) whereas with pure functions I don't really know. I agree that this should be private _get... . Why double?

File: models.py [hunk 6]
```diff
@@ -280,19 +280,38 @@ def __get_distance_weights(d, w1, w0, sigma):
...
+    if ce_loss is None:
+        ce_loss = torch.nn.CrossEntropyLoss()
+        ce_target = dice_target
+    return dice_weight * dice_loss(output, dice_target) + ce_weight * ce_loss(output, ce_target)
+
+
+def dice_loss(output, target):
+    dice_numerator = 0
+    dice_denominator = 0
+    for class_nr in range(1, int(target.max()) + 1):
```
Comment (jakubczakon) [inline models.py]: @apyskir this assumes that 0 is background. Could case trouble when we encode/preprocess stuff differently

File: models.py [hunk 7]
```diff
@@ -280,19 +280,38 @@ def __get_distance_weights(d, w1, w0, sigma):
...
+def mixed_dice_cross_entropy_loss(output, target, dice_weight, ce_weight, ce_loss=None):
+    dice_target = target[:, 0, :, :].long()
+    ce_target = target
+    if ce_loss is None:
+        ce_loss = torch.nn.CrossEntropyLoss()
+        ce_target = dice_target
+    return dice_weight * dice_loss(output, dice_target) + ce_weight * ce_loss(output, ce_target)
+
+
+def dice_loss(output, target):
```
Comment (jakubczakon) [inline models.py]: @apyskir I think it is always a good idea to have some eps. But apparently having smooth could also help. As descibed herehttps://github.com/pytorch/pytorch/issues/1249 with this being a code snipped for smooth (no eps):

```
def dice_loss(input, target):
    smooth = 1.

    iflat = input.view(-1)
    tflat = target.view(-1)
    intersection = (iflat * tflat).sum()
    
    return 1 - ((2. * intersection + smooth) /
              (iflat.sum() + tflat.sum() + smooth))
```

File: models.py [hunk 8]
```diff
@@ -280,19 +280,38 @@ def __get_distance_weights(d, w1, w0, sigma):
...
+    dice_numerator = 0
+    dice_denominator = 0
+    for class_nr in range(1, int(target.max()) + 1):
+        class_target = (target == class_nr)
```
Comment (jakubczakon) [inline models.py]: I would really like to use the regular one-category dice inside of this loop. That makes for less debuging

File: models.py [hunk 9]
```diff
@@ -280,19 +280,38 @@ def __get_distance_weights(d, w1, w0, sigma):
...
+        class_target.data = class_target.data.float()
+        dice_numerator += (class_target * output[:, class_nr, :, :]).sum()
+        dice_denominator += (class_target + output[:, class_nr, :, :]).sum()
```
Comment (jakubczakon) [inline models.py]: @apyskir it should be the same but I would go with class_target.sum()+output...sum() for the readability.

File: models.py (hunk 8 above)
Comment (jakubczakon) [reply models.py]: @apyskir also I don't think that we are doing any softmax/sigmoid after the output goes through the network. We should perform it here (which one?) . I think this may be causing the nan by the way the fact that activations are so high sometimes that they shoot the loss to the sky.

File: loaders.py (hunk 1 above)
Comment (apyskir) [reply loaders.py]: @jakubczakon Actually the function get_datagen() here is identical to the one defined in ImageSegmentationLoaderBasic, so I removed it, because it does not need to be overriden.
Yes, it was causing some trouble, I think during validation on epoch end.
Hmm actually to calculate `validation_sum` you need to do `mask_transform`, despite it is not train mode. `train_mode` may be a little bit confusing.

File: models.py (hunk 2 above)
Comment (apyskir) [reply models.py]: @jakubczakon This solution is temporary (in principle, in practice maybe it is not), because w1 should be part of a target. So building w1 is really temporary. I can move it to `get_loss_params`, but when building w1 is fixed, it will need to be changed again.

File: models.py (hunk 3 above)
Comment (apyskir) [reply models.py]: @jakubczakon This formulation of dice_loss allows you to use it with just target matrix. For weighted cross entropy you need target matrix, distances and sizes and there is no possibility, that weighted cross entropy will work with simple target matrix. I can implement retrieving target matrix from `target` variable inside dice_loss, but then it will not be reusable to cases where you don't have these auxiliary targets.

File: models.py (hunk 4 above)
Comment (apyskir) [reply models.py]: @jakubczakon OK, fine.

File: models.py (hunk 5 above)
Comment (apyskir) [reply models.py]: @jakubczakon This may be my misunderstanding of the idea of putting single or double underscore before function name. If we want it to be private, it should be with double, right? Otherwise it's just a name with single underscore, but in practice it is public.
OK, I can change it.

File: models.py (hunk 6 above)
Comment (apyskir) [reply models.py]: @jakubczakon Yes, I think that this is a natural and justified assumption. Why would you mark background with different label? In case of binary models (like in DSB) you also made this assumption.

File: models.py [hunk 10]
```diff
@@ -263,36 +266,56 @@ def multiclass_weighted_segmentation_loss(output, target, w0, sigma):
...
         w0 = w0.cuda()
         sigma = sigma.cuda()
-    return {'w0': w0, 'sigma': sigma}
+        C = C.cuda()
+    return {'w0': w0, 'sigma': sigma, 'C': C}
+
+
+def mixed_dice_cross_entropy_loss(output, target, dice_weight, ce_weight, ce_loss=None):
+    dice_target = target[:, 0, :, :].long()
+    ce_target = target
```
[the remark this answers is not in the document]
Comment (apyskir) [reply models.py]: @jakubczakon I am not sure. You can pass any function here as a parameter, but if you don't, the default cross entropy loss is nn.CrossEntropyLoss(), which is not weighted. So naming the target weighted_cross_entropy_target in my opinion is not right. I would keep ce_target or change it to cross_entropy_target for readability.

File: models.py [hunk 11]
```diff
@@ -263,36 +271,52 @@ def multiclass_weighted_segmentation_loss(output, target, w0, sigma):
...
+    dice_target = target[:, 0, :, :].long()
+    cross_entropy_target = target
+    if cross_entropy_loss is None:
+        cross_entropy_loss = torch.nn.CrossEntropyLoss()
+        cross_entropy_target = dice_target
+    return dice_weight * dice_loss(output, dice_target, smooth) + cross_entropy_weight * cross_entropy_loss(output,
+                                                                                                    cross_entropy_target)
+
+
+def dice_loss(output, target, smooth=0):
```
Comment (jakubczakon) [inline models.py]: @apyskir I really like it like this

File: steps/pytorch/validation.py [hunk 12]
```diff
@@ -6,13 +6,16 @@
 
 
 class DiceLoss(nn.Module):
-    def __init__(self):
+    def __init__(self, smooth=0, eps = 1e-7):
```
Comment (jakubczakon) [inline steps/pytorch/validation.py]: @apyskir @taraspiotr 
Can we go for Softmax in the first iteration? 
I do want to experiment with Sigmoid here but let's start simple (without mixing softmax in CE and sigmoid in DICE). Am I missing something?
